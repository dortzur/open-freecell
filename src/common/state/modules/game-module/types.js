export const START_GAME = 'START_GAME';
export const PERFORM_MOVE = 'PERFORM_MOVE';
