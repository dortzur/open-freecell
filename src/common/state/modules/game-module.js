import { srand } from '../../utils';
import { deck } from '../../models/deck';
import { normalize } from 'normalizr';
import schema from '../schema';

export const START_GAME = 'START_GAME';

export default (state = {}, action) => {
  switch (action.type) {
    case START_GAME: {
      const data = normalize({ deck: action.payload.gameDeck }, schema);
      return { ...state, deck: data.result.deck };
    }
    default: {
      return state;
    }
  }
};

export const startGame = (gameNumber) => {
  const gameDeck = [...deck];
  const rand = srand(gameNumber);
  for (let i = gameDeck.length - 1; i > 0; i--) {
    const r = rand() % (i + 1);
    const card = gameDeck[r];
    gameDeck[r] = gameDeck[i];
    gameDeck[i] = card;
  }
  gameDeck.reverse();
  return { type: START_GAME, payload: { gameDeck } };
};

const getCard = (state, id) => state.entities.cards[id];
