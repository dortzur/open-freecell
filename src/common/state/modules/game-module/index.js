import { normalize } from 'normalizr';
import schema from '../../schema';
import produce from 'immer';
import * as types from './types';

const initialState = {
  foundations: [[], [], [], []],
  cells: [[], [], [], []],
  tableau: [[], [], [], [], [], [], [], []],
};

export default (state = initialState, action) => {
  switch (action.type) {
    case types.START_GAME: {
      const { gameDeck } = action.payload;

      const data = normalize({ deck: gameDeck }, schema);
      const deck = data.result.deck;

      state.tableau = produce(state.tableau, (tableau) =>
        deck.reduce((tableau, card, index) => {
          tableau[index % 8].push(card);
          return tableau;
        }, tableau)
      );

      return { ...state };
    }
    default: {
      return state;
    }
  }
};
