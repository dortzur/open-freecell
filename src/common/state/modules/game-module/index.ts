import { GAME_STATUS } from '../../../utils/consts';
import * as boardTypes from '../board-module/types';
import { AnyAction } from 'redux';

const getInitialState = () => ({
  status: GAME_STATUS.NOT_STARTED,
  number: null,
});


export default (state = getInitialState(), action: AnyAction) => {
  switch (action.type) {
    case boardTypes.START_GAME: {
      const { gameNumber } = action.payload;
      return { ...state, status: GAME_STATUS.STARTED, number: gameNumber };
    }
    default: {
      return state;
    }
  }
};

