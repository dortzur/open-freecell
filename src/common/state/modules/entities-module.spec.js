import { startGame } from './game-module';
import entitiesReducer from './entities-module';

describe('legacy', () => {
  it('creates game 100', () => {
    const actionResult = startGame(100);
    expect(entitiesReducer(undefined, actionResult)).toMatchSnapshot();
  });
});
