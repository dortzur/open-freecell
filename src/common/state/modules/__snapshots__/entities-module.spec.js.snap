// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`legacy creates game 100 1`] = `
Object {
  "cards": Object {
    "2C": Object {
      "color": "black",
      "id": "2C",
      "rank": "2",
      "suit": "Clubs",
      "value": 2,
    },
    "2D": Object {
      "color": "red",
      "id": "2D",
      "rank": "2",
      "suit": "Diamonds",
      "value": 2,
    },
    "2H": Object {
      "color": "red",
      "id": "2H",
      "rank": "2",
      "suit": "Hearts",
      "value": 2,
    },
    "2S": Object {
      "color": "black",
      "id": "2S",
      "rank": "2",
      "suit": "Spades",
      "value": 2,
    },
    "3C": Object {
      "color": "black",
      "id": "3C",
      "rank": "3",
      "suit": "Clubs",
      "value": 3,
    },
    "3D": Object {
      "color": "red",
      "id": "3D",
      "rank": "3",
      "suit": "Diamonds",
      "value": 3,
    },
    "3H": Object {
      "color": "red",
      "id": "3H",
      "rank": "3",
      "suit": "Hearts",
      "value": 3,
    },
    "3S": Object {
      "color": "black",
      "id": "3S",
      "rank": "3",
      "suit": "Spades",
      "value": 3,
    },
    "4C": Object {
      "color": "black",
      "id": "4C",
      "rank": "4",
      "suit": "Clubs",
      "value": 4,
    },
    "4D": Object {
      "color": "red",
      "id": "4D",
      "rank": "4",
      "suit": "Diamonds",
      "value": 4,
    },
    "4H": Object {
      "color": "red",
      "id": "4H",
      "rank": "4",
      "suit": "Hearts",
      "value": 4,
    },
    "4S": Object {
      "color": "black",
      "id": "4S",
      "rank": "4",
      "suit": "Spades",
      "value": 4,
    },
    "5C": Object {
      "color": "black",
      "id": "5C",
      "rank": "5",
      "suit": "Clubs",
      "value": 5,
    },
    "5D": Object {
      "color": "red",
      "id": "5D",
      "rank": "5",
      "suit": "Diamonds",
      "value": 5,
    },
    "5H": Object {
      "color": "red",
      "id": "5H",
      "rank": "5",
      "suit": "Hearts",
      "value": 5,
    },
    "5S": Object {
      "color": "black",
      "id": "5S",
      "rank": "5",
      "suit": "Spades",
      "value": 5,
    },
    "6C": Object {
      "color": "black",
      "id": "6C",
      "rank": "6",
      "suit": "Clubs",
      "value": 6,
    },
    "6D": Object {
      "color": "red",
      "id": "6D",
      "rank": "6",
      "suit": "Diamonds",
      "value": 6,
    },
    "6H": Object {
      "color": "red",
      "id": "6H",
      "rank": "6",
      "suit": "Hearts",
      "value": 6,
    },
    "6S": Object {
      "color": "black",
      "id": "6S",
      "rank": "6",
      "suit": "Spades",
      "value": 6,
    },
    "7C": Object {
      "color": "black",
      "id": "7C",
      "rank": "7",
      "suit": "Clubs",
      "value": 7,
    },
    "7D": Object {
      "color": "red",
      "id": "7D",
      "rank": "7",
      "suit": "Diamonds",
      "value": 7,
    },
    "7H": Object {
      "color": "red",
      "id": "7H",
      "rank": "7",
      "suit": "Hearts",
      "value": 7,
    },
    "7S": Object {
      "color": "black",
      "id": "7S",
      "rank": "7",
      "suit": "Spades",
      "value": 7,
    },
    "8C": Object {
      "color": "black",
      "id": "8C",
      "rank": "8",
      "suit": "Clubs",
      "value": 8,
    },
    "8D": Object {
      "color": "red",
      "id": "8D",
      "rank": "8",
      "suit": "Diamonds",
      "value": 8,
    },
    "8H": Object {
      "color": "red",
      "id": "8H",
      "rank": "8",
      "suit": "Hearts",
      "value": 8,
    },
    "8S": Object {
      "color": "black",
      "id": "8S",
      "rank": "8",
      "suit": "Spades",
      "value": 8,
    },
    "9C": Object {
      "color": "black",
      "id": "9C",
      "rank": "9",
      "suit": "Clubs",
      "value": 9,
    },
    "9D": Object {
      "color": "red",
      "id": "9D",
      "rank": "9",
      "suit": "Diamonds",
      "value": 9,
    },
    "9H": Object {
      "color": "red",
      "id": "9H",
      "rank": "9",
      "suit": "Hearts",
      "value": 9,
    },
    "9S": Object {
      "color": "black",
      "id": "9S",
      "rank": "9",
      "suit": "Spades",
      "value": 9,
    },
    "AC": Object {
      "color": "black",
      "id": "AC",
      "rank": "Ace",
      "suit": "Clubs",
      "value": 1,
    },
    "AD": Object {
      "color": "red",
      "id": "AD",
      "rank": "Ace",
      "suit": "Diamonds",
      "value": 1,
    },
    "AH": Object {
      "color": "red",
      "id": "AH",
      "rank": "Ace",
      "suit": "Hearts",
      "value": 1,
    },
    "AS": Object {
      "color": "black",
      "id": "AS",
      "rank": "Ace",
      "suit": "Spades",
      "value": 1,
    },
    "JC": Object {
      "color": "black",
      "id": "JC",
      "rank": "Jack",
      "suit": "Clubs",
      "value": 11,
    },
    "JD": Object {
      "color": "red",
      "id": "JD",
      "rank": "Jack",
      "suit": "Diamonds",
      "value": 11,
    },
    "JH": Object {
      "color": "red",
      "id": "JH",
      "rank": "Jack",
      "suit": "Hearts",
      "value": 11,
    },
    "JS": Object {
      "color": "black",
      "id": "JS",
      "rank": "Jack",
      "suit": "Spades",
      "value": 11,
    },
    "KC": Object {
      "color": "black",
      "id": "KC",
      "rank": "King",
      "suit": "Clubs",
      "value": 13,
    },
    "KD": Object {
      "color": "red",
      "id": "KD",
      "rank": "King",
      "suit": "Diamonds",
      "value": 13,
    },
    "KH": Object {
      "color": "red",
      "id": "KH",
      "rank": "King",
      "suit": "Hearts",
      "value": 13,
    },
    "KS": Object {
      "color": "black",
      "id": "KS",
      "rank": "King",
      "suit": "Spades",
      "value": 13,
    },
    "QC": Object {
      "color": "black",
      "id": "QC",
      "rank": "Queen",
      "suit": "Clubs",
      "value": 12,
    },
    "QD": Object {
      "color": "red",
      "id": "QD",
      "rank": "Queen",
      "suit": "Diamonds",
      "value": 12,
    },
    "QH": Object {
      "color": "red",
      "id": "QH",
      "rank": "Queen",
      "suit": "Hearts",
      "value": 12,
    },
    "QS": Object {
      "color": "black",
      "id": "QS",
      "rank": "Queen",
      "suit": "Spades",
      "value": 12,
    },
    "TC": Object {
      "color": "black",
      "id": "TC",
      "rank": "10",
      "suit": "Clubs",
      "value": 10,
    },
    "TD": Object {
      "color": "red",
      "id": "TD",
      "rank": "10",
      "suit": "Diamonds",
      "value": 10,
    },
    "TH": Object {
      "color": "red",
      "id": "TH",
      "rank": "10",
      "suit": "Hearts",
      "value": 10,
    },
    "TS": Object {
      "color": "black",
      "id": "TS",
      "rank": "10",
      "suit": "Spades",
      "value": 10,
    },
  },
}
`;
